### PR TITLE
Add DebugValue for function param regardless of scope

### PR DESCRIPTION
**Pull request by jaebaek**
`DebugInfoManager::AddDebugValueIfVarDeclIsVisible()` adds
OpenCL.DebugInfo.100 DebugValue from DebugDeclare only when the
DebugDeclare is visible to the given scope. It helps us correctly
handle a reference variable e.g.,

```
{ // scope #1.
  int foo = /* init */;
  { // scope #2.
    int& bar = foo;
    ...
```

in the above code, we must not propagate DebugValue of `int& bar` for
store instructions in the `scope #1` because it is alive only in
the `scope #2`.

We have an exception: If the given DebugDeclare is used for a function
parameter, `DebugInfoManager::AddDebugValueIfVarDeclIsVisible()` has
to always add DebugValue instruction regardless
of the scope. It is because the initializer (store instruction) for
the function parameter can be out of the function parameter's scope
(the function) in particular when the function was inlined.

Without this change, the function parameter value information always
disappears whenever we run the function inlining pass and
`DebugInfoManager::AddDebugValueIfVarDeclIsVisible()`.

### CODE REVIEW

File: source/opt/debug_info_manager.cpp [hunk 1]
```diff
@@ -523,7 +545,10 @@ void DebugInfoManager::AddDebugValueIfVarDeclIsVisible(
                                    kDebugValueOperandLocalVariableIndex),
                                value_id, 0, index_id, insert_before);
     assert(added_dbg_value != nullptr);
-    added_dbg_value->UpdateDebugInfoFrom(scope_and_line);
+    if (is_function_param)
+      added_dbg_value->UpdateDebugInfoFrom(dbg_decl_or_val);
+    else
+      added_dbg_value->UpdateDebugInfoFrom(scope_and_line);
```
Comment (kuhar) [inline source/opt/debug_info_manager.cpp:551]:
```suggestion
    added_dbg_value->UpdateDebugInfoFrom(is_function_param ? dbg_decl_or_val : scope_and_line);
```

**Review by kuhar — COMMENTED** [review]
Can you surround the example with ` ``` ` (markdown)? Same with `#1` and `#2` -- otherwise they get misparsed by github.

**Comment by jaebaek** [comment]
Done!

**Review by kuhar — APPROVED** [review]
Thanks, looks good to me.

**Comment by kuhar** [comment]
Are the segfaults with the two windows CI bots unrelated to this change?

**Comment by ben-clayton** [comment]
Not very familiar with this code, but looks okay to me. Thanks for the fix @jaebaek!

**Comment by jaebaek** [comment]
Thank you for the code review. In terms of the CI bot failures, I think it is related to the googletest updates. I believe #3922 temporarily fixes it. Let me rebase it and submit it.

**Pull request merged**